icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1131,10 +1131,13 @@ private static int getShortestSubtagLength(String localeID) {
      * @stable ICU 3.0
      */
     public static String getName(String localeID){
-        String tmpLocaleID;
+        String tmpLocaleID = localeID;
         // Convert BCP47 id if necessary
         if (localeID != null && !localeID.contains("@") && getShortestSubtagLength(localeID) == 1) {
-            tmpLocaleID = forLanguageTag(localeID).getName();
+            if (!localeID.toLowerCase().startsWith("x-") || localeID.indexOf('_') < 0) {
```

Comment:
Different condition from C++: There you test "if contains underscore and not..." but here you test "if ... *or does not contain underscore*". Pretty sure this is wrong :-)